### ISSUE TITLE
[AMD][KDA] Port chunk-parallel prefill from Triton to Gluon on GFX950

### ISSUE THREAD

**Issue by raikonenfnu** [raikonenfnu/tokenspeed#41]
## Summary

Track the GFX950 port of Kimi Delta Attention (KDA) chunk-parallel prefill from the original native Triton implementation to Triton Gluon.

The first implementation is in draft PR lightseekorg/tokenspeed#858. It removes AMD prefill's runtime dependency on Flash Linear Attention (FLA), registers an AMD GFX950 Gluon specialization through the vendor-neutral `tokenspeed-kernel` boundary, and preserves the canonical K-major state ABI `[N, H, K, V]`.

- Original Triton implementation: [`kda_chunk.py`](https://github.com/lightseekorg/tokenspeed/blob/0f68676069141857a605b8805c13131d9f53e901/tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/triton/kda_chunk.py)
- Gluon port: [`kda_prefill_gfx950.py`](https://github.com/raikonenfnu/tokenspeed/blob/raikonen/kda_clean/tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/attention/gluon/kda_prefill_gfx950.py)
- Draft integration PR: lightseekorg/tokenspeed#858

## Original algorithm

The serial KDA recurrence updates a per-head, per-key-channel state matrix `H`:

```text
q = l2norm(q) / sqrt(K)
k = l2norm(k)
g = safe_gate(A_log, g_raw + dt_bias)
beta = sigmoid(beta_logits)

H_tilde = exp(g_t) * H
delta   = beta_t * (v_t - k_t @ H_tilde)
H       = H_tilde + k_t^T @ delta
o_t     = q_t @ H
```

Serial token recurrence is appropriate for decode but underutilizes the GPU during long prefills. The Triton implementation applies the gated delta-rule WY factorization over chunks:

```text
bg = chunk_cumsum(g)
Kd = Kn * exp(bg)
Ki = Kn * exp(-bg)
Qd = scale * Qn * exp(bg)

A = tril(-diag(beta) * Kd * Ki^T, -1)
T = (I - A)^-1
u = T * (beta * V)
W = T * (beta * Kd)

# sequential only across chunks
v_new = u - W * H

# parallel within each chunk
o = Qd * H + tril(Qd * Ki^T, 0) * v_new
H = exp(bg_last) * H + (Kn * exp(bg_last - bg))^T * v_new
```

Chunk-local exponentials use a subchunk reference so the exponent differences remain bounded and do not overflow FP32.

## Triton-to-Gluon mapping

| Algorithm stage | Original Triton | GFX950 Gluon | Notes |
|---|---|---|---|
| Q/K normalization and gate/beta preparation | shared L2 norm plus `_kda_prepare_gate_beta_kernel` | `_prepare_qk_gate_beta_kernel` | Fused in Gluon to reduce launches and intermediate traffic. |
| Chunk-local gate prefix sum | `chunk_local_cumsum_vector` | same portable helper | Resets at every packed-sequence boundary. |
| Strict-lower WY matrix | `_kkt_vector_fwd_kernel` | `_kkt_vector_fwd_kernel` | Same stabilized `Kd * Ki^T` math using CDNA4 MFMA. |
| Triangular inverse | `solve_tril` | same portable helper | Produces BF16 `T` for subsequent MFMA operands. |
| WY transformed values and keys | `_wu_vector_fwd_kernel` | `_wu_vector_fwd_kernel` | Computes `u` and `W`. |
| Sequential inter-chunk recurrence | `_state_scan_fwd_kernel` | `_state_scan_fwd_kernel` | Keeps each `[K,V]` state tile in FP32, emits checkpoints and `v_new`. |
| Parallel chunk output | `_output_fwd_kernel` | `_output_fwd_kernel` | Computes both inter- and causal intra-chunk contributions. |
| Gate application | `_gate_apply_kernel` | `_gate_apply_kernel` | Materializes bounded `beta*Kd` and `beta*V`. |

The Gluon `AMDMFMALayout` and operand layouts are register/workgroup layouts only. They do not change the externally visible state ABI. State loads and stores use `key * V + value`, i.e. canonical `[N,H,K,V]` storage.

## Intentional implementation differences

- The Gluon kernel is currently specialized for Kimi K3 dimensions `K=V=128` on GFX950.
- Chunk size is 64 and subchunk size is fixed at 16. This matches the production `lower_bound=-5` Triton path. The original Triton implementation could choose a larger subchunk for other bounds.
- Normalized Q/K intermediates are stored in BF16 in the current Gluon implementation instead of FP32. MFMA accumulation and persistent state remain FP32.
- The original native Triton wrapper accepted a legacy `[N,H,V,K]` state and transposed internally. The current kernel API directly consumes and returns the repository's canonical `[N,H,K,V]` state.
- Cumsum, packed-sequence indexing, and triangular solve remain portable Triton helpers; the KDA-specific matrix and recurrence stages are Gluon.

## Correctness validation

- Packed variable-length sequences `[65, 17]` match a direct token-by-token PyTorch recurrence for both `lower_bound=-5` and the unbounded-gate path.
- Focused KDA, dispatch, and setup suite: **13 passed**.
- Gluon vs FLA state cosine similarity: approximately **0.9999996**.
- Kimi K3 AIME26 canary, first eight prompts with high reasoning, temperature 1.0, top-p 0.95, and 32K cap: **8/8**.
- 4K/1K end-to-end completes with no coherence or dispatch failures.

## Current performance and remaining work

This is deliberately a draft because the first Gluon port is not yet performance-equivalent to FLA:

| Measurement | Gluon port | FLA / prior baseline |
|---|---:|---:|
| 4K operator prefill | 3.93 ms | 1.47 ms |
| 128K operator prefill | 127.3 ms | 41.3 ms |
| 4K/1K mean TTFT | 681.5 ms | 543.0 ms |
| 4K/1K mean TPOT | 23.88 ms | 23.99 ms |

Decode TPOT is unchanged, isolating the regression to prefill. The largest current costs are the parallel output stage and sequential state scan. Follow-up work should profile and tune their MFMA tiling, checkpoint traffic, launch decomposition, and opportunities to fuse or pipeline state/output work without changing the algorithm or cache ABI.

## Acceptance criteria

- Preserve the KDA recurrence and canonical `[N,H,K,V]` state ABI.
- Maintain packed-variable-length coherence and K3 accuracy.
- Match or beat the portable FLA implementation at both 4K and approximately 128K prefill lengths.
- Avoid a required FLA dependency for AMD installations.
- Keep runtime dispatch vendor-neutral through `tokenspeed-kernel`.
